handler-ssl-ocsp/src/main/java/io/netty/handler/ssl/ocsp/DefaultConnectionChannelFactory.java

```diff
@@ -0,0 +1,103 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl.ocsp;
+
+import io.netty.channel.ChannelFactory;
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.Epoll;
+import io.netty.channel.epoll.EpollDatagramChannel;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.epoll.EpollSocketChannel;
+import io.netty.channel.kqueue.KQueue;
+import io.netty.channel.kqueue.KQueueDatagramChannel;
+import io.netty.channel.kqueue.KQueueEventLoopGroup;
+import io.netty.channel.kqueue.KQueueSocketChannel;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.DatagramChannel;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioDatagramChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.resolver.dns.DnsNameResolver;
+import io.netty.resolver.dns.DnsNameResolverBuilder;
+
+final class DefaultConnectionChannelFactory implements ConnectionChannelFactory {
+
+    static final DefaultConnectionChannelFactory INSTANCE = new DefaultConnectionChannelFactory();
+
+    private EventLoop eventLoop;
+    private DnsNameResolver dnsNameResolver;
+
+    @Override
+    public EventLoop eventLoop() {
+        if (eventLoop == null) {
+            if (KQueue.isAvailable()) {
+                eventLoop = new KQueueEventLoopGroup(1).next();
+            } else if (Epoll.isAvailable()) {
+                eventLoop = new EpollEventLoopGroup(1).next();
+            } else {
+                eventLoop = new NioEventLoopGroup(1).next();
+            }
+        }
+        return eventLoop;
+    }
+
+    @Override
+    public SocketChannel socketChannel() {
+        if (KQueue.isAvailable()) {
+            return new KQueueSocketChannel();
+        } else if (Epoll.isAvailable()) {
+            return new EpollSocketChannel();
+        } else {
+            return new NioSocketChannel();
+        }
+    }
+
+    @Override
+    public DnsNameResolver dnsNameResolver() {
+        if (dnsNameResolver == null) {
+            DnsNameResolverBuilder resolver = new DnsNameResolverBuilder()
+                    .eventLoop(eventLoop())
+                    .channelFactory(new ChannelFactory<DatagramChannel>() {
+                        @Override
+                        public DatagramChannel newChannel() {
+                            if (KQueue.isAvailable()) {
+                                return new KQueueDatagramChannel();
+                            } else if (Epoll.isAvailable()) {
+                                return new EpollDatagramChannel();
+                            } else {
+                                return new NioDatagramChannel();
+                            }
```

Comment:
If we pluck the event loop from the context, then we're missing a good way to ensure this channel is compatible with the given event loop. Hmm…